Cell and material instance counting for PyAPI

wbinventor opened the pull request:
This PR introduces new features in the PyAPI to count the number of instances of each cell and material in a geometry as I suggested in #797. This is intended to help support @cjosey with his development of OpenDeplete.

The new `Geometry.count_cell_instances()` and `Geometry.count_material_instances()` methods can be called to recursively count the number of instances of cells/materials in a geometry. The final counts are stored in the new `Cell.num_instances` and `Material.num_instances` properties, respectively. An example of how this new code be used and tested (*i.e.*, in one of our Python example input files) could be as follows:

```python
geometry.count_cell_instances()
geometry.count_material_instances()
for cell in geometry.get_all_cells().values():
    print(cell.id, cell.name, cell.num_instances)
for material in geometry.get_all_materials().values():
    print(material.id, material.name, material.num_instances)
```

@cjosey do you want to take a look at this and see if you can use it as a way to generically determine the number of distribmats you need to define for OpenDeplete?

cjosey commented:
I just tested this with my example geometries and it works well.  None of the code to my knowledge looks problematic.  I'm happy.

It is currently incompatible with how I've done ``example_geometry.py`` though.  In that, all fuel pins are mapped to the same cell, so I need some auxiliary information to determine what fuel goes in what fuel pin.  This implicitly requires knowledge of the number of fuel pins.  In the BEAVRS case, though, I am pretty sure that all different fuel types are assigned to different cells, so that would be an ideal use-case.

I did have a question though.  In the example code I based my geometry on, it had a fill for the universes, and yet this code reports that the universe has no materials to fill it with.  Has the need to fill a universe changed with time?

wbinventor commented:
@cjosey would the issue you describe resolve itself if you required that each depletable cell started out with a unique fuel material?

I'm not sure about your question regarding universe fills, is this specific to my code, or about changes made to OpenMC more generally?

cjosey commented:
I do not believe that this is an issue anymore.  The example code I based mine off of no longer assigns a material to the root universe.  

And yes, requiring each cell to contain a unique initial condition would solve the issue entirely.  It would also solve my concerns about extreme scale problems.  The initial condition would be defined by the PyAPI without distribmats, my code would load it, perform distribmats (in parallel, what have you), and go on its way.

In these circumstances, for example, BEAVRS would not require any modification at all other than setting material.burnable and material.volume.

wbinventor commented:
Do these new features solve the issues we've discussed @cjosey, or do you also need the front-end API to query the distribcell/mat paths?

cjosey commented:
I believe it solves all the issues.  Correct me if I'm wrong, but the paths would just be used to find where in the geometry a material corresponds with, right?  I could see a few uses for that:
1. Plotting reaction rates / number densities in the core.
2. Setting distribmats in a non-one cell, one initial condition fashion.

The first would be really cool, but I'd have to put together a big huge geometry plotter code.  For the second, I vaguely remember hearing at one point that the paths are nonobvious to the point where a user would have difficulty leveraging them.

paulromano commented on openmc/cell.py:
```diff
@@ -90,6 +90,10 @@ class Cell(object):
+    num_instances : int
+        The number of instances of this cell throughout the geometry. This
+        property is initialized by calling the
+        :meth:`Geometry.count_cell_instances()` method.
```
No parentheses needed inside that `:meth:` directive.

paulromano commented on openmc/geometry.py:
```diff
@@ -455,3 +455,105 @@ def get_lattices_by_name(self, name, case_sensitive=False, matching=False):
+
+            # If material-filled, we are finished with all levels
+            if cell.fill_type == 'material':
+                continue
```
I would just remove this part of the if block.

paulromano commented on openmc/geometry.py:
```diff
@@ -455,3 +455,105 @@ def get_lattices_by_name(self, name, case_sensitive=False, matching=False):
+                valid_indices = latt.indices
+
+                # Count instances in each Universe in the Lattice
+                for index in valid_indices:
```
remove line above and make this `for index in latt.indices`

paulromano replied:
Or even `for index in cell.fill.indices`

wbinventor replied:
Good call

paulromano commented on openmc/geometry.py:
```diff
@@ -455,3 +455,105 @@ def get_lattices_by_name(self, name, case_sensitive=False, matching=False):
         lattices.sort(key=lambda x: x.id)
         return lattices
+
+    def count_cell_instances(self, universe=None):
```
Seems like maybe this should be a method of `Universe`. `Geometry` could have its own version that resets the instance counts and then calls `self.root_universe.count_cell_instances()`. What do you think?

wbinventor replied:
Sure, I like that design, it's just as we implemented this algorithm in its original form in OpenMOC. Though I think `count_cell_instances()` should be a private `Universe` class method since we don't want users randomly calling it on any universe they wish.

paulromano replied:
Yeah, good idea on making that method private

paulromano commented on openmc/lattice.py:
```diff
@@ -950,26 +957,24 @@ def universes(self, universes):
         elif isinstance(self._universes[0][0][0], openmc.Universe):
-            n_dims = 3
-
+            pass
```
Correct me if I'm wrong -- this entire `if` block looks superfluous since the call the `cv.check_iterable_type` above will achieve the same purpose.

wbinventor replied:
You're absolutely right. This must be code from a long time a go in a galaxy far, far away without a `checkvalue` module.

paulromano commented on openmc/material.py:
```diff
@@ -71,6 +71,10 @@ class Material(object):
+    num_instances : int
+        The number of instances of this cell throughout the geometry. This
+        property is initialized by calling the
+        :meth:`Geometry.count_cell_instances()` method.
```
Remove parentheses

paulromano replied:
Also, should be count_material_instances

paulromano commented on openmc/material.py:
```diff
@@ -71,6 +71,10 @@ class Material(object):
         mol.  For example, UO2 with 3 nuclides will have an average molar mass
         of 270 / 3 = 90 g / mol.
+    num_instances : int
+        The number of instances of this cell throughout the geometry. This
```
...of this material...

paulromano commented on openmc/cell.py:
```diff
@@ -344,6 +353,11 @@ def distribcell_paths(self, distribcell_paths):
         self._distribcell_paths = distribcell_paths
 
+    @num_instances.setter
+    def num_instances(self, num_instances):
```
I would argue that we shouldn't even have a setter for `num_instances` since we wouldn't want a user setting it themselves. Can we just have the `count` methods directly set the private attribute?

wbinventor replied:
Haha I debated this with myself. I'm fine with the `Geometry` directly incrementing the "private" `_num_instances` property. That's why this is written in Python after all :-)

paulromano commented on openmc/cell.py:
```diff
@@ -228,6 +233,10 @@ def distribcell_paths(self):
 
+    @property
+    def num_instances(self):
+        return self._num_instances
```
Maybe we should raise an exception here if `_num_instances is None` and tell the user they need to run the `count_cell_instances` method.

wbinventor replied:
Sure sounds good

paulromano commented on openmc/lattice.py:
```diff
@@ -922,6 +922,13 @@ def indices(self):
+        if isinstance(self.universes[0][0], openmc.Universe):
+            return 2
+        else:
+            return 3
```
`return 2 if isinstance(self.universes[0][0], openmc.Universe) else 3`

paulromano replied:
Also, thanks for adding this missing property!

wbinventor commented:
@cjosey yes you are right about the paths. And you are also right about the paths being somewhat non-obvious. If they aren't needed for any of the code you have already written, I propose to leave them out of this PR.

paulromano commented:
"somewhat non-obvious" :) @wbinventor I was actually able to generate paths with a few slight modifications to the code in this PR, so it's won't be that bad. I do want to sync it up with what the Fortran side spits out (in the summary file). So, if you're cool with it, I can submit a PR after this one is merged that adds support to show paths.

On that topic, the paths that Fortran spits out are "somewhat non-obvious" themselves because they don't really tell you what's a cell and a what's a universe. For example, this is what I get by reading in a summary file generated in test_distribmat:
```python
In [21]: cells[11].distribcell_paths
Out[21]: 
['0->101->101(1,1,1)->11->11',
 '0->101->101(1,2,1)->11->11',
 '0->101->101(2,1,1)->11->11',
 '0->101->101(2,2,1)->11->11']
```
I'm thinking it might be better to have strings that look like `'u0->c101->l101(1,1,1)->u11'`. What do you think? (Notice I've left off the last 11 because that is the cell itself). The other option is that the paths could be lists of `Universe`, `Lattice`, and `Cell` objects, i.e. `[<Universe 0>, <Cell 101>, (<Lattice 101>, (1,1,1)), <Universe 11>]`.

wbinventor commented:
@paulromano I've addressed all of your comments. I'm certainly fine if you want to submit your code to generate paths as a follow-on PR - it would certainly be nice to have. If nothing else, the algorithm is probably much easier to follow in your Python implementation than in Fortran, so it will serve to document the scheme :-)

I agree that the path strings in the "summary.h5" file aren't very useful in and of themselves. I would be fine annotating each ID with a `'m'`, `'c'`, `'u'`, or `'l'` prefix for material, cell, universe or lattice, but would probably refrain from adding to much extraneous information to minimize the memory footprint. Perhaps it would be more useful to have the logic to parse the paths in the PyAPI to either (1) build a list of the `Material`, `Cell`, `Universe` and `Lattice` objects for a particular path on-the-fly and/or (2) further annotate the path strings on-the-fly to make them more human-readable. What do you think?

paulromano commented on openmc/universe.py:
```diff
@@ -519,3 +519,53 @@ def create_xml_subelement(self, xml_element):
+
+            # If material-filled, we are finished with all levels
+            if cell.fill_type == 'material':
+                cell.fill._num_instances += 1
```
You also need to account for 'distribmat' fill_type

paulromano replied:
Thinking about this a little more, it is a little complicated. Each time you recurse to the lowest level, you would only want to increment the instance count for the corresponding material, but there's no way to know how many times you've reached this cell (and hence which material in `cell.fill` you should be incrementing), that is unless you count cell and material instances in the same method, in which case you could do something like
```Python
elif cell.fill_type == 'distribmat':
    cell.fill[cell._num_instances]._num_instances += 1
```

paulromano commented on openmc/material.py:
```diff
@@ -196,6 +201,14 @@ def depletable(self):
+    def num_instances(self):
+        if self._num_instances is None:
+            raise ValueError('The number of cell instances is unknown. Call '
+                             'the Geometry.count_cell_instances() method.')
```
Change cell to material in this message

paulromano commented:
@wbinventor I have a working implementation for paths, but it changes much of the code you have here, mostly because I had to combine counting cells and materials (this is the only way to get materials correct). I've also changed the name of the method on `Geometry` from `count_*_instances` to `determine_paths` (open to other suggestions). So, given how different it looks, I'm going to just suggest closing this PR and I will open a new one with the paths implementation that builds on the commits from this branch. Are you ok with that?

wbinventor commented:
@paulromano I'll close this PR and review yours when you are ready. I have no real preference for the naming convention, I vote for whatever you think would be most intuitive to a user.
